### PR TITLE
Mempool cache

### PR DESCRIPTION
**Pull request by melekes**

### CODE REVIEW

File: docs/transactional-semantics.rst [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+Tendermint"<./specification/using-tendermint.html#broadcast-api>`__ we
+discussed different API endpoints for sending transactions and
+differences between them.
+
+What we have not yet covered is transactional semantics.
+
+When you send a transaction using one of the available methods, it
+first goes to the mempool. Currently, it does not provide strong
+guarantees like "if the transaction were accepted, it would be
+eventually included in a block (given CheckTx passes)."
```
Comment (ebuchman) [inline docs/transactional-semantics.rst]: we should mention that if you want the guarantees, you can just use `_commit` (which will either be committed or return an error, unlike _sync), but then you have to deal with higher latency. This is the tradeoff.

File: mempool/mempool.go [hunk 2]
```diff
@@ -497,5 +497,6 @@ func (cache *txCache) Push(tx types.Tx) bool {
 func (cache *txCache) Remove(tx types.Tx) {
 	cache.mtx.Lock()
 	delete(cache.map_, string(tx))
+	// so the list grows unboundedly? shouldn't we remove tx from the list here too?
```
Comment (ebuchman) [inline mempool/mempool.go]: the list is capped in `Push`, where we remove from the map and list. So it's ok if we don't remove from the list in Remove, but I don't really see why we wouldn't (maybe performance ?).

File: docs/transactional-semantics.rst (hunk 1 above)
Comment (melekes) [reply docs/transactional-semantics.rst]: I assumed people have read "Using tendermint Broadcast API" where we explain the difference. But I guess it's worth repeating. Will do

File: mempool/mempool.go (hunk 2 above)
Comment (melekes) [reply mempool/mempool.go]: assuming an attacker creates txs that do not pass CheckTx, list grows unboundedly.

File: docs/transactional-semantics.rst (hunk 1 above)
Comment (ebuchman) [reply docs/transactional-semantics.rst]: Can we not consolidate this doc with that one ?

File: mempool/mempool.go (hunk 2 above)
Comment (ebuchman) [reply mempool/mempool.go]: I don't understand why another transaction got added after CheckTx fails.

File: docs/transactional-semantics.rst (hunk 1 above)
Comment (melekes) [reply docs/transactional-semantics.rst]: I thought this one will be more technical (mentioning things like WAL, ...)

File: mempool/mempool.go (hunk 2 above)
Comment (melekes) [reply mempool/mempool.go]: I was wrong.

**Pull request merged**